cmake/Modules/ValkeySetup.cmake

```diff
@@ -0,0 +1,349 @@
+function (valkey_get_distro_name DISTRO_NAME)
+    if (LINUX AND NOT APPLE)
+        execute_process(
+            COMMAND /bin/bash "-c" "cat /etc/os-release |grep ^ID=|cut -d = -f 2"
+            OUTPUT_VARIABLE _OUT_VAR
+            OUTPUT_STRIP_TRAILING_WHITESPACE)
+        # clean the output
+        string(REPLACE "\"" "" _OUT_VAR "${_OUT_VAR}")
+        string(REPLACE "." "" _OUT_VAR "${_OUT_VAR}")
+        set(${DISTRO_NAME}
+            "${_OUT_VAR}"
+            PARENT_SCOPE)
+    elseif (APPLE)
+        set(${DISTRO_NAME}
+            "darwin"
+            PARENT_SCOPE)
+    elseif (IS_FREEBSD)
+        set(${DISTRO_NAME}
+            "freebsd"
+            PARENT_SCOPE)
+    else ()
+        set(${DISTRO_NAME}
+            "unknown"
+            PARENT_SCOPE)
+    endif ()
+endfunction ()
+
+# Determine if we are building in Release or Debug mode
+if (CMAKE_BUILD_TYPE MATCHES Debug OR CMAKE_BUILD_TYPE MATCHES DebugFull)
+    set(VALKEY_DEBUG_BUILD 1)
+    message(STATUS "Building in debug mode")
+else ()
+    set(VALKEY_DEBUG_BUILD 0)
+    message(STATUS "Building in release mode")
+endif ()
+
+# ----------------------------------------------------
+# Helper functions - end
+# ----------------------------------------------------
+
+# ----------------------------------------------------
+# Build options (allocator, tls, rdma et al)
+# ----------------------------------------------------
+
+if (NOT WITH_MALLOC)
+    if (APPLE)
+        set(WITH_MALLOC "libc")
+    elseif (UNIX)
+        set(WITH_MALLOC "jemalloc")
+    endif ()
+endif ()
+
+# User may pass different allocator library. Using -DWITH_MALLOC=<libname>, make sure it is a valid value
+if (WITH_MALLOC)
+    if ("${WITH_MALLOC}" STREQUAL "jemalloc")
+        set(MALLOC_LIB "jemalloc")
+        add_valkey_server_compiler_options("-DUSE_JEMALLOC")
+        set(USE_JEMALLOC 1)
+    elseif ("${WITH_MALLOC}" STREQUAL "libc")
+        set(MALLOC_LIB "libc")
+    elseif ("${WITH_MALLOC}" STREQUAL "tcmalloc")
+        set(MALLOC_LIB "tcmalloc")
+        add_valkey_server_compiler_options("-DUSE_TCMALLOC")
+    elseif ("${WITH_MALLOC}" STREQUAL "tcmalloc_minimal")
+        set(MALLOC_LIB "tcmalloc_minimal")
+        add_valkey_server_compiler_options("-DUSE_TCMALLOC")
+    else ()
+        message(FATAL_ERROR "WITH_MALLOC can be one of: jemalloc, libc, tcmalloc or tcmalloc_minimal")
+    endif ()
+endif ()
+
+message(STATUS "Using ${MALLOC_LIB}")
+
+# TLS support
+if (WITH_TLS)
+    find_package(OpenSSL REQUIRED)
+    message(STATUS "OpenSSL include dir: ${OPENSSL_INCLUDE_DIR}")
+    message(STATUS "OpenSSL libraries: ${OPENSSL_LIBRARIES}")
+
+    include_directories(${OPENSSL_INCLUDE_DIR})
+    add_valkey_server_compiler_options("-DUSE_OPENSSL=1")
+    add_valkey_server_compiler_options("-DBUILD_TLS_MODULE=0")
+    set(USE_TLS 1)
+endif ()
+
+if (WITH_RDMA)
+    # RDMA support (Linux only)
+    if (LINUX AND NOT APPLE)
+        add_valkey_server_compiler_options("-DUSE_RDMA=2")
```

Comment:
Is it possible to add a helper function to parse `module/yes/no` for both TLS and RDMA? even for QUIC in the future.
Then we have `-DWITH_TLS=yes`, `-DWITH_TLS=module` and `-DWITH_TLS=no` rather than `-DWITH_TLS=1`.

Reply:
Same question here - do we get real benefit from making these components true modules? The build time optionality is there. The run-time optionality is lost but I am curious to know if this is indeed needed in a production environment?

Reply:
The module style connection driver proves the well abstracted framework. Haha!
Sure, builtin connection driver is fine to me, too! I agree that there is no obvious benefit from module components.

Reply:
From what I see in the Makefile, RDMA is either supported as module or not at all

```make
$(error "RDMA is only supported as module (BUILD_RDMA=module), or disabled (BUILD_RDMA=no)")
```

Reply:
I'm working on `BUILD_RDMA=yes`, I plan to support it this week.

Reply:
Added support for:

```bash
cmake .. -DWITH_TLS=module # optionally you can pass here: yes|no|1|0
```

The same can be done with RDMA:

```bash
cmake .. -DWITH_RDMA=module # optionally you can pass here: yes|no|1|0
```

While you can still pass `1` or `yes` to`-DWITH_RDMA` options, internally, it will be disabled as we only support module mode for RDMA